planar graph recognition

l0gun0v opened the pull request.

krzysztof-turowski committed:
Add common greedy vertex coloring heuristics (#3)

* Add RS and LF coloring algorithms
* Let vertex coloring algorithms inherit from Algorithm
* Add SL greedy coloring heuristic
* Add SLF coloring algorithm
* Add GIS vertex coloring algorithm
* Add list of algorithms

krzysztof-turowski committed:
Fix gtest dependency

krzysztof-turowski committed:
Add perfect graphs recognition algorithm (#4)

* Add first version of BFS with predicate
* Add GraphTools::toComplement
* Port Adrian Siwiec algorithm for recognition of perfect graphs
* Migrate hasJewel to main code
* Migrate hasT1 to main code
* Migrate hasPyramid to main code
* Migrate hasT2 to main code
* Migrate hasT3 to main code
* Add isT3 method
* Migrate hasNearCleanerOddHole to main code
* Remove temporary checks
* Provide alias for Bitset
* Fix linter issues
* Update submodules

krzysztof-turowski committed:
King-Rao-Tarjan maximum flow algorithm (#6)

* Add all maxflow files
* Enable maxflow KRT algorithm
* Remove dependency on KRTGraph
* Move the logic to KRTMaxFlow class
* Convert 1-based to 0-based notation
* Implement benchmark test from DIMACS file
* Update README.md

krzysztof-turowski committed:
Groetschel-Lovasz-Schrijver perfect graph coloring algorithm (#7)

* First upload of perfect graph coloring algorithm
* Fix extern C error
* Fix compilation
* Benchmark tested up to 8 vertices
* Migrate getTheta to NetworKit::Graph
* Second part of migration
* Third part of migration
* Fourth part of migration
* Fix GraphTools
* Final migration
* Remove theta.cpp
* Remove commons.cpp
* Fixes for clang++ compilation
* Final cleanups

krzysztof-turowski committed:
Cleanups and fixes (#10)

* Cleanup of files and examples
* Add small graphs
* Fix G6 writer
* Update submodules

PiotrKubaty committed:
Implementation of exact algorithms for minimum dominating set (#8)

* Rooij-Bodlaender mds algorithm
* Fomin-Kratsch-Woeginger algorithm
* Schiermeyer algorithm
* Refactor: split RooijBodlanderMDS algorithm into 3 algorithms (Grandoni, FGK, Rooij-Bodlaender) with inheritance and templates
* Add benchmark for dominating sets
* Cleanup

zglapa committed:
Add enumeration vertex coloring algorithms (#9)

* Add Brown's ordinary enumeration vertex coloring
* Add Christofides and Brelaz enumeration vertex colorings
* Add Kormans enumeration coloring algorithm
* Refactor algorithms to work with vertex numbers relative to ordering
* Fix interchange returning bad solution on some occasions
* Add benchmarks
* Apply review comments (#5)
* Remove DRVertexQueue in favour of Aux::BucketPQ
* Add thesis pdf and sources

krzysztof-turowski committed:
Update .gitattributes

krzysztof-turowski committed:
Clean dominating set and vertex coloring exponential algorithms (#11)

* Rename theses
* Update dependencies
* Refactor BranchAndReduce set cover files
* Clean dominating set algorithms
* Remove ExhaustiveMDS
* Migrate dominating_set from vector to set
* Fix readers and writers
* Clean vertex coloring algorithms

Onchao committed:
Implementation of exact algorithms for maximum independent set (#5)

* Add bruteforce independent set algorithm.
* Mis1 algorithm added and reimplemented with NetworKit::Graph
* Mis5 added
* Mis 1,3,4,5 completed
* Neighbor functions refactorization
* Case degree 2 for Mis2
* Mis2 refactored but its bugged
* Tests for all graphs with small size
* Further fixes and benchmark test
* Parametrize benchmarks
* Mis2 fixed and added case for degree 3 vertices
* Added missing Mis3 degree 3 case and fixed again
* Mis2 case with vertex with degree >= 6 added
* Mis2 done
* Measure and Conquer algorithm but folding bugged
* Measure and Conquer algorithm done
* Class hierarchy fixed and redundant test removed
* EdgeSet comparator is now static and max degree for a graph function removed
* BruteForce and utility functions optimized
* Neighbor2 function optimized
* Redundant includes removed
* check() function added and getIndependentSet() returns set now
* DFS removed and added final changes except algorithms description/authors
* Add thesis files
* Clean code and update README

krzysztof-turowski committed:
Add minimum spanning tree finding and verification algorithms  (#12)

* Commit original code
* Implement proxy for all MST algorithms
* Implement Prim's MST algorithm
* Migrate Boruvka algorithm
* Start migrating KKT algorithm
* Provide theoretically better heaps for Prim
* Migrate FibonacciHeap to vector
* Migrate PairingHeap to vector
* Migrate BinomialHeap to vector
* Implement tests for heaps
* Migrate MST verification
* Finalize KKT MST algorithm migration
* Final cleanups
* Update the list of algorithms

krzysztof-turowski committed:
Remove submodule dependendy on gtest and boost (#20)

* Update submodules
* Remove submodule dependency on gtest and boost

l0gun0v committed:
Hopcroft Tarjan

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,303 @@
+                            (S.top())->flip();
+                        }
+                        if(b->left_interlace(S)){
+                            delete b;
```
In general, please avoid using pointers and new/delete unless it's really necessary

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,303 @@
+       
+
+        //test planarity
+        std::vector<NetworKit::node> parent(N, INT_MAX);
```
Again, please use `NetworKit::none`

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,303 @@
+        reorder(g, parent, dfsnum, order);
+    
+        std::vector<int> alpha(g.numberOfEdges());
+        {
```
Why did you introduce local block?

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,303 @@
+        return true;
+    }
+
+    bool HopcroftTarjan::planar() {
```
Consider changing output type to PlanarGraphRecognition::State

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    void HopcroftTarjan::make_biconnected_graph(NetworKit::Graph &g) {
+        // make it connected
+        NetworKit::node u = 0;
+        std::vector<bool> reached(N);
+        for(auto v: g.nodeRange()){
+            if(!reached[v]) {
+                dfs(g, v, reached);
+                if(u != v) {
+                   g.addEdge(u, v);
+                   g.addEdge(v, u);
+                }
+            }
+        }
+        // make it biconnected
+        fill(reached.begin(), reached.end(), false);
+        std::vector<NetworKit::node> lowpt(N);
+        std::vector<NetworKit::node> parent(N, INT_MAX);
+        std::vector<NetworKit::node> dfsnum(N);
+        int dfs_count = 0;
```
Please replace finding biconnected components with https://networkit.github.io/dev-docs/cpp_api/classNetworKit_1_1BiconnectedComponents.html#_CPPv4N9NetworKit21BiconnectedComponentsE

In general it's worthwhile to look at the NetworKit C++ docs to see if one can avoid reimplementing lower-level algorithms. E.g. simple DFS is there too and AFAIK you can pass a lambda there, so that you should be able to build `lowpt` etc.

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,303 @@
+
+    HopcroftTarjan::HopcroftTarjan(NetworKit::Graph &graph, bool embedding): IPlanarGraphRecognition(graph, embedding){ }
+
+    void HopcroftTarjan::dfs(NetworKit::Graph &g,
+                            NetworKit::node v, 
+                            std::vector<bool> & reached
+                            )
+    {
+        reached[v] = true;
+        for(auto u: g.neighborRange(v)){
+            if(!reached[u]){
+                dfs(g, u, reached);
+            }
+        }
+    }
```
There's a DFS in NetworKit library, no need to reimplement it.

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+
+class HopcroftTarjan : public IPlanarGraphRecognition {
+   public:
+      HopcroftTarjan::HopcroftTarjan(NetworKit::Graph &graph, bool embedding);
```
Please reuse the constructor - see e.g. ExactVertexColoring and its subclasses for details

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+
+};
+
+class HopcroftTarjan : public IPlanarGraphRecognition {
```
I know it's long, but probably we should go with HopcroftTarjanPlanarGraphRecognition - I wouldn't bet a dime that they didn't develop other algorithms together.

krzysztof-turowski replied:
Please also add an analogue of this docstring for perfect graph recognition (substituting source etc.)
```
/**
 * @ingroup recognition
 * The class for recognition of perfect graphs procedure from
 * Chudnovsky, Cornuejols, Liu, Seymour, Vuskovic, "Recognizing Berge graphs".
 *
 */
```

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+     
+
+      void dfs(NetworKit::Graph &g, NetworKit::node v, std::vector<bool> & reached);
+      void dfs_in_make_biconnected_graph(NetworKit::Graph &g, NetworKit::node v, 
+             int &dfs_count, std::vector<bool> &reached, std::vector<NetworKit::node> &dfsnum,
+             std::vector<NetworKit::node> &lowpt, std::vector<NetworKit::node> &parent);
```
Here you have a nice way (almost) of indenting long functions, please refactor all longer headings to fit the scheme:
```
void dfs_in_make_biconnected_graph(
    NetworKit::Graph &g, NetworKit::node v, int &dfs_count, std::vector<bool> &reached, 
    std::vector<NetworKit::node> &dfsnum, std::vector<NetworKit::node> &lowpt,
    std::vector<NetworKit::node> &parent);
```

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+ *
+ *  Created on: 24.03.2024
+ *      Author: Dzianis Lahunou
+ *      Ported by: Krzysztof Turowski (krzysztof.szymon.turowski@gmail.com)
```
You don't need my name here and elsewhere, you will commit the code to this repo by yourself :)

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+ * The Interface for class for recognition of planar graphs p.
+ *
+ */
+class IPlanarGraphRecognition : public NetworKit::Algorithm {
```
Just `PlanarGraphRecognition`, it's an abstract class, not an interface.

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+     * @return true if the graph is planar, false otherwise.
+     */
+   bool isPlanar() const;
+
+
```
Please leave only one empty line

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return true if the graph is planar, false otherwise.
+     */
+   bool isPlanar() const;
```
Consider `getState` as in `PerfectGraphRecognition` instead

krzysztof-turowski commented on test/testPlanarGraphRecognition.cpp:
```diff
@@ -0,0 +1,33 @@
+    algorithm.run();
+   
+    auto is_planar = algorithm.isPlanar(); 
+    std::cout << is_planar << "\n";
```
Please go with std::endl;

krzysztof-turowski commented on cpp/recognition/CMakeLists.txt:
```diff
@@ -4,4 +4,7 @@ koala_add_module(recognition
    perfect/Pyramids.cpp
    perfect/NearCleaners.cpp
+
+   planar/IPlanarGraphRecognition.cpp
```
Please imitate PerfectGraphRecognition structure, that is:
- move `planar/IPlanarGraphRecognition.cpp` -> `PlanarGraphRecognition.cpp`
- consider moving `planar/HopcroftTarjan.cpp` -> `HopcroftTarjanPlanarGraphRecognition.cpp`
- you don't need an empty line between

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+         ha.splice(ha.end(), Ratt); Ratt.splice(Ratt.end(), Latt); Latt.splice(Latt.end(), ha);
+         he.splice(he.end(), Rseg); Rseg.splice(Rseg.end(), Lseg); Lseg.splice(Lseg.end(), he);
+      }
+      int head_of_Latt() {
+         return Latt.front();
+      }
+      bool empty_Latt() {
+         return Latt.empty();
+      }
+      int head_of_Ratt() {
+         return Ratt.front();
+      }
+      bool empty_Ratt() {
+         return Ratt.empty();
+      }
```
I guess since you use these functions only locally, you can remove them, they save next to nothing.

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+         return Ratt.empty();
+      }
+      bool left_interlace(std::stack<Block*> &s){
+         if(!s.empty() && !((s.top())->empty_Latt()) && Latt.back() < (s.top())->head_of_Latt()){
+            return true;
+         }
+         return false;
```
`return !s.empty() && !((s.top())->empty_Latt()) && Latt.back() < (s.top())->head_of_Latt()`?

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+      }
+
+      bool right_interlace(std::stack<Block*> &s){
+         if(!s.empty() && !((s.top())->empty_Ratt()) && Latt.back() < (s.top())->head_of_Ratt()){
+            return true;
+         }
+         return false;
```
See above

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+ *
+ */
+
+class Block {
```
Why don't you move this to HopcroftTarjan.cpp? It's a local structure, no need to expose it to anyone using hpp header.

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+
+namespace Koala {
+
+const int left = 1;
+const int right = 2;
```
Please use UPPERCASE - and as with Block, consider moving to the respective cpp file

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
Since hpp files are pretty short, please merge with PlanarGraphRecognition.hpp (and move Block out of it to the cpp file).

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+
+#include <optional>
+#include <vector>
+#include <limits.h>
```
If you get rid of INT_MAX, I guess it'll be superfluous,

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+
+/**
+ * @ingroup recognition
+ * The Interface for class for recognition of planar graphs p.
```
Maybe "The abstract class for recognition of planar graphs"?

krzysztof-turowski commented on include/recognition/planar/IPlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Return the embbending found by the algorithm.
+     *
+     * @return pointer to embedding if the graph is planar, nullptr otherwise.
```
Prefer returning embedding by value, e.g. `std::optional<Embedding>` if you have to have nullpointer

krzysztof-turowski reviewed (changes requested):
Thank you very much, very nice code, so don't get discouraged by the number of comments - the contribution is of really good quality and I hope it won't be a problem for you to fix those issues.

krzysztof-turowski commented on include/recognition/planar/HopcroftTarjan.hpp:
```diff
@@ -0,0 +1,148 @@
+
+const int left = 1;
+const int right = 2;
+/**
+ * @ingroup recognition
+ * The Interface for class for recognition of planar graphs p.
+ *
+ */
```
Remove this, I guess it's copypasted

PiJayson committed:
Implement several minimum and maximum cut algorithms (#24)

* max-cut: new three algorithms
* min-cut: new four algorithms
* added two more algorithms. fixed GoemansWilliamson
* fixed pending code style issues
* added benchmark for max-cut. code refactor
* min-cut: added benchmark
* added bachelor paper

fixikmila committed:
Implement cograph recognition algorithms (#21)

Implement BretscherCorneilHabibPaulCographRecognition, CorneilStewartPerl, and a sequential version of Dahlhaus algorithm.

aSmallTyphoon committed:
Implementation of algorithms on cographs (#23)

rkilar committed:
Maximum weight matching algorithms implementation (#25)

* Initial implementations for Gabow and Galil Micali Gabow
* Work in progress augmentation handling
* Add lazy path augmentation
* Finished MicaliGabow
* Started Micali Vazirani algorithm
* Change MV algorithm
* Changes to MV algorithm
* MV fixed
* Minor MV cleanup
* Final MV cleanup
* Scaling algorithm done
* Minor scaling speedup
* Minor updates
* Split find min implementation
* Clean up
* Minor fixes to maximum matching algorithms
* Add comments, clean up unused code
* Add comments and documentation
* Add comments for the scaling algorithm
* Update comments and names for the scaling algorithm
* Made names more consistent
* Change some priority queues, fix naming
* Better naming
* Slight optimizations
* Optimize the Micali Gabow algorithm
* Remove my scripts from gitignore
* Added tests for maximum matching algorithms
* Cleanup code
* Rename algorithm to include all authors
* Change name of heap
* Update names
* Remove debug code
* Fix linter errors in header files, small fixes and cleanup
* Minor cleanup
* Fix memory leaks and corner cases
* Fix memory leaks in Galil Micali Gabow
* Added thesis draft
* Small Gabow scaling algorithm optimazation
* Change names to be more consistent with thesis
* Fix Gabow scaling memory leaks
* Fix errors in the maximum matching text
* Split matching data structures into separate files
* Clean up priority queue files
* Added perfect matching options
* Add tests, fix linting errors
* Added greedy optimization option for perfect matching
* Improve benchmark output
* Add experimental results
* Optimize Galil Micali Gabow algorithm
* Further optimizations
* Optimize the array priority queue
* Update experimental results summary
* Initial text corrections
* Corrections draft
* More corrections
* Remove shortened definitions
* Final corrections
* Final corrections
* Update figures
* Final final corrections
* Bibliography typo fixes
* Revert some optimization to align with text
* Clean up code
* Remove debug code
* Update networkit
* Apply style changes
* Update networkit

l0gun0v committed:
merge

l0gun0v committed:
fix

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+            Lseg.push_back(e);
+            Latt.splice(Latt.end(), A);
+        }
+        ~Block() {}
+        void flip() {
```
Nit: please provide one empty line between functions

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+
+        void add_to_Att(std::list<int> &Att, int dfsnum_w0,
+                        std::vector<int> &alpha) {
+            if (!Ratt.empty() && this->Ratt.front() > dfsnum_w0) flip();
```
Nit: please go for
```
if (...) {
   flip();
}
```
even for one-line commands.

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+
+    class Block {
+    private:
+        std::list<int> Latt, Ratt;
+        std::list<NetworKit::WeightedEdge> Lseg, Rseg;
```
Suggestion: please use full names instead of "att" and "seg", it'll be easier to read your code to outsiders.

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+    HopcroftTarjan::HopcroftTarjan(NetworKit::Graph &graph, bool embedding)
+            : PlanarGraphRecognition(graph, embedding) {}
+
+    void HopcroftTarjan::dfs(NetworKit::Graph &g, NetworKit::node v,
+                             std::vector<bool> &reached) {
+        reached[v] = true;
+        for (auto u : g.neighborRange(v)) {
+            if (!reached[u]) {
+                dfs(g, u, reached);
+            }
+        }
+    }
```
Please use instead https://networkit.github.io/dev-docs/cpp_api/function_DFS_8hpp_1a4fad57413d253911caaded68e09a9100.html#_CPPv4I0EN9NetworKit9Traversal7DFSfromEvRK5Graph4node1L

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    void HopcroftTarjan::make_biconnected_graph(NetworKit::Graph &g) {
+        // make it connected
+        NetworKit::node u = 0;
+        std::vector<bool> reached(N);
+        for (auto v : g.nodeRange()) {
+            if (!reached[v]) {
+                dfs(g, v, reached);
+                if (u != v) {
+                    g.addEdge(u, v);
+                    g.addEdge(v, u);
+                }
+            }
+        }
+        // make it biconnected
+        fill(reached.begin(), reached.end(), false);
+        std::vector<NetworKit::node> lowpt(N);
+        std::vector<NetworKit::node> parent(N, NetworKit::none);
+        std::vector<NetworKit::node> dfsnum(N);
+        int dfs_count = 0;
+
+        dfs_in_make_biconnected_graph(g, u, dfs_count, reached, dfsnum, lowpt,
+                                      parent);
+    }
+
```
Please use https://networkit.github.io/dev-docs/cpp_api/classNetworKit_1_1BiconnectedComponents.html methods instead

krzysztof-turowski commented on cpp/recognition/planar/PlanarGraphRecognition.cpp:
```diff
@@ -0,0 +1,26 @@
+ *
+ *  Created on: 24.03.2024
+ *      Author: Dzianis Lahunou
+ *      Ported by: Krzysztof Turowski (krzysztof.szymon.turowski@gmail.com)
```
This line is not needed, you write it directly to this library :)

krzysztof-turowski commented on include/recognition/planar/PlanarGraphRecognition.hpp:
```diff
@@ -0,0 +1,113 @@
+ *
+ *  Created on: 24.03.2024
+ *      Author: Dzianis Lahunou
+ *      Ported by: Krzysztof Turowski (krzysztof.szymon.turowski@gmail.com)
```
As above

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+                                      parent);
+    }
+
+    void HopcroftTarjan::dfs_in_reorder(NetworKit::Graph &g, NetworKit::node v,
+                                        std::vector<NetworKit::Edge> &Add,
+                                        int &dfs_count, std::vector<bool> &reached,
+                                        std::vector<NetworKit::node> &dfsnum,
+                                        std::vector<NetworKit::node> &lowpt,
+                                        std::vector<NetworKit::node> &lowpt2,
+                                        std::vector<NetworKit::node> &parent) {
+        dfsnum[v] = dfs_count++;
```
Again, it is advisable to go with DFSFrom method from NetworKit library - unless you have a really good reason not to do so.

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    bool HopcroftTarjan::strong_planar(NetworKit::Graph &g, NetworKit::Edge e,
+                                       std::list<int> &Att, std::vector<int> &alpha,
+                                       std::vector<NetworKit::node> &dfsnum,
+                                       std::vector<NetworKit::node> &parent) {
```
Maybe https://networkit.github.io/dev-docs/cpp_api/classNetworKit_1_1StronglyConnectedComponents.html might be useful

krzysztof-turowski commented on test/CMakeLists.txt:
```diff
@@ -11,3 +10,5 @@ koala_make_test(test_minimum_spanning_tree testMinimumSpanningTree.cpp)
 koala_make_test(test_minimum_cut testMinCut.cpp)
 koala_make_test(test_maximum_cut testMaxCut.cpp)
+koala_make_test(test_perfect_graph_recognition testPerfectGraphRecognition.cpp)
+koala_make_test(test_planar_graph_recognition testPlanarGraphRecognition.cpp)
```
Please add empty line at the end

krzysztof-turowski commented on test/testPlanarGraphRecognition.cpp:
```diff
@@ -0,0 +1,56 @@
+                4, {{0, 1}, {0, 2}, {1, 3}, {2, 3}}, Koala::PlanarGraphRecognition::State::PLANAR},
+        PlanarGraphRecognitionParameters{
+                5, {{0, 1}, {1, 2}, {2, 3}, {3, 4}}, Koala::PlanarGraphRecognition::State::PLANAR},
+        PlanarGraphRecognitionParameters{5,
+                                         {{0, 1},
+                                          {0, 2},
+                                          {0, 3},
+                                          {0, 4},
+                                          {1, 2},
+                                          {1, 3},
+                                          {1, 4},
+                                          {2, 3},
+                                          {2, 4},
+                                          {3, 4}},
+                                         Koala::PlanarGraphRecognition::State::NOT_PLANAR},
+        PlanarGraphRecognitionParameters{6,
+                                         {{0, 3},
+                                          {0, 4},
+                                          {0, 5},
+                                          {1, 3},
+                                          {1, 4},
+                                          {1, 5},
+                                          {2, 3},
+                                          {2, 4},
+                                          {2, 5}},
+                                         Koala::PlanarGraphRecognition::State::NOT_PLANAR}));
```
Please write these examples more concisely, like the example in lines 32-33 (just above!) in this file

krzysztof-turowski commented on benchmark/benchmarkPlanarGraphs.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <sys/stat.h>
+#include <sys/types.h>
```
Please remove these headers

krzysztof-turowski commented on benchmark/benchmarkPlanarGraphs.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <cassert>
+#include <fstream>
+#include <io/G6GraphReader.hpp>
+#include <iostream>
+#include <map>
+#include <recognition/planar/PlanarGraphRecognition.hpp>
```
Please split these headers into two groups: native C++ includes + NetworKit includes (see other benchmark files for reference)

krzysztof-turowski commented on cpp/recognition/planar/HopcroftTarjan.cpp:
```diff
@@ -0,0 +1,358 @@
+        if (embedding) {
+            // TODO embedding
+        }
+        return PlanarGraphRecognition::State::PLANAR;;
```
Remove one unnecessary semicolon

krzysztof-turowski reviewed (changes requested):
Thank you very much, nice piece of work - just please fix few style issues + replace your own algorithms (e.g. DFS, SCC) by the ones already implemented in NetworKit... unless there's a really strong rationale for not doing so

l0gun0v committed:
test

l0gun0v committed:
lr

l0gun0v committed:
fix

l0gun0v committed:
fix style issues

bartefix committed:
Push-relabel, Malhotra-Kumar-Maheshwari, Boykov-Kolmogorov maximum flow algorithms (#41)

* Maxflow
* Fix complexity in MKM
* Added text
* Refactor

---------

Co-authored-by: bartefix <bart@DESKTOP-5T2C9KL.localdomain>

289Adam289 committed:
Frederickson and Heizinger et al. planar graph single source shortest path algorithms (#39)

* Temporary build config
* Initial sssp structure
* Implement simplified FCS algorith and extract boost helper functions
* suitable r-division wip
* Section 3 wip
* Section 3 first draft
* create suitable r-division quickly
* Add Topology Based Heap structure and boundry vertices preprocesing
* Topology based heap logic - first full algorithm draft
* connected subsets debuged
* debuged makeSuitableGraphDivision
* findSuitableRDivision first working version
* add simple mop up function
* first draft fully debuged
* Add level 2 division
* Frederickson fully debugged both levels
* Extract r-division logic
* Henzinger planarSSSP implementation
* format code
* clean up comments
* Add text
* Rename text subdirectories
* Fix style issues
* Align PlanarSSSP with NetworKit SSSP
* Fix file structure and method names
* Fix types and remove boost hash function
* Fix spelling

jkukowski committed:
Implement various priority queue data structures (#37)

* Implement van Emde Boas tree
* Add thesis chapter structure and bibliography
* van Emde Boas tree chapter
* Make van Emde Boas tree generic
* Implement X-fast trie and Y-fast trie
* Fix X-fast trie bugs, add X-fast trie chapter
* Refactor Y-fast trie, add Treap utility implementation and Y-fast trie chapter
* Weak heap implementation and chapter
* Add implementation and chapter for skew heap and rank-pairing heap
* Add introduction and testing chapters, apply text fixes, add benchmarks
* Refactor most text chapters, improve benchmark charts
* Improve benchmark legend text and image
* Adjust PriorityQueue interface, adhere variable naming to repo conventions, fix cpplint issues

l0gun0v committed:
fix
